pkg/extensions/tbtc/tbtc.go

```diff
@@ -761,23 +790,58 @@ func (t *tbtc) watchKeepClosed(
 }
 
 func (t *tbtc) shouldMonitorDeposit(depositAddress string) bool {
+	t.monitoredDepositsCache.Sweep()
+	t.notMonitoredDepositsCache.Sweep()
+
+	if t.monitoredDepositsCache.Has(depositAddress) {
+		return true
+	}
+
+	if t.notMonitoredDepositsCache.Has(depositAddress) {
+		return false
+	}
+
+	if _, err := t.getSignerIndex(depositAddress); err != nil {
+		t.notMonitoredDepositsCache.Add(depositAddress)
+		return false
+	}
+
+	t.monitoredDepositsCache.Add(depositAddress)
+	return true
+}
+
+func (t *tbtc) getSignerIndex(depositAddress string) (int, error) {
 	keepAddress, err := t.chain.KeepAddress(depositAddress)
 	if err != nil {
-		return false
+		return 0, err
 	}
 
 	members, err := t.chain.GetMembers(common.HexToAddress(keepAddress))
 	if err != nil {
-		return false
+		return 0, err
 	}
 
-	for _, member := range members {
+	for index, member := range members {
 		if member == t.chain.Address() {
-			return true
+			return index, nil
 		}
 	}
 
-	return false
+	return 0, fmt.Errorf(
```

Comment:
Instead of returning an error, I'd probably return `-1, nil`. Then, inside `shouldMonitorDeposit` we can decide whether we should not monitor the deposit or if we do not want we should be doing that because of an error during the check.